kpatch-build/create-diff-object.c

```diff
@@ -760,11 +760,39 @@ static void kpatch_compare_symbols(struct list_head *symlist)
 	}
 }
 
+#define CORRELATE_ELEMENTS(e1, e2)				\
+do {								\
+	e1->twin = e2;						\
+	e2->twin = e1;						\
+	/* set initial status, might change */			\
+	e1->status = e2->status = SAME;				\
+} while (0)
+
+static void kpatch_process_section_correlation(struct section *sec)
+{
+	if (!sec->twin)
+		return;
+
+	if (is_rela_section(sec)) {
+		CORRELATE_ELEMENTS(sec->base, sec->twin->base);
+		sec = sec->base;
+	} else if (sec->rela) {
+		CORRELATE_ELEMENTS(sec->rela, sec->twin->rela);
+	}
+
+	if (sec->secsym)
+		CORRELATE_ELEMENTS(sec->secsym, sec->twin->secsym);
+	if (sec->sym)
+		CORRELATE_ELEMENTS(sec->sym, sec->twin->sym);
+}
+
 static void kpatch_correlate_sections(struct list_head *seclist1, struct list_head *seclist2)
 {
 	struct section *sec1, *sec2;
 
 	list_for_each_entry(sec1, seclist1, list) {
+		if (sec1->twin)
```

Comment:
The `kpatch_process_section()` might already have done the correlation for an associated rela/base section. In which case, looking for another element with a matching mangled name can match to a different object (since the correct one will be already correlated, i.e. `sec2->twin != NULL`, we can't match to it again).

Reply:
Do you mean `kpatch_correlate_sections()`? I don't see how that would happen since it goes through each list only once.

Reply:
No, sorry I meant, `kpatch_process_section_correlation()` which I introduce in this commit.

Maybe its introduction is being a bit overzealous, but taking multiple objects  with matching mangled names into consideration (which was the original motivation for @jpoimboe 's change), I don't know how e.g. we guarantee we would avoid the following scenario:
orig:
- sym_a -> bundled into sec_a

patched:
- sym_b -> bundled into sec_b
- sym_c -> bundled into sec_c

If both sym_b and sym_c have names matching sym_a through mangling, we could end up with sym_a being correlated with sym_b but sec_a being correlated with sec_c.
Same thing with rela/base sections.

So basically I introduced the `kpatch_process_section_correlation()` to avoid that. But the consequence is that when you correlate a section, you might correlate another one (and a symbol) before having treated in in the list_foreach. If that happens, it needs to be skipped.

Reply:
Ok, that makes more sense now. Not sure how correct is this approach but I don't have a better solution on hand.